Clothing mod expansion

nphyx opened the pull request:
#### Summary
```SUMMARY: Features "expands clothing mod functionality"```

#### Purpose of change
Compared to other equipment mods, bionics/mutagens, and vehicle customization, modifying clothing is a little boring and limited. The intent is to give them a broader range of capabilities.

A PR in progress plans to overhaul the tailoring system in general but didn't look like it was doing anything with mods yet, and might be related: https://github.com/CleverRaven/Cataclysm-DDA/pull/33806#issue-314168388

#### Describe the solution
**More mod capabilities**:
- add additional item flags while the mod is active (without changing the item's own `item_tags`)
- suppress flags while the mod is active
- change an item's coverage value
- have variable durations
- have variable difficulty levels
- explicitly set the base number of input items (still scaled by volume)
- disable volume scaling for input items (e.g. if you want to consume only one input no matter the target item's size)

**Fixes**:
- tweaked skill roll and difficulty equation to allow advanced tailors to more easily add multiple mods 
  - it was nearly impossible to add more than 2 before, regardless of skill level
  - it should work out about the same as before for low/unskilled tailors
- clamped final coverage calculation to ensure coverage stayed within 0-100
- made all other properties minimum 0 (no more negative warmth/etc caused by mods)
- unified mod compatibility checks
- general tidying of sew_advanced_actor::use
    
**Allows mods to be more specifically targeted**, enabling generic mods while preventing incoherent combinations:
- can require certain flags or exclude certain flags
- can require or exclude certain covered body parts
- can require a minimum and maximum coverage value
    
**Some UI work on clothing mod tools (iuse_actor.cpp)**:
- item values that aren't changed by a mod are no longer displayed in the description (reduces clutter)
- mods that are completely incompatible are not listed (clutter)
  - still shows mods that could be applied if you had the materials or met other conditions
- if the mod applies or removes flags, lists them in the description
- can't apply a mod to a worn item that changes flags (causes bugs)
- shows a time estimate for how long the mod will take to complete

**New content**:
- added an embroidery kit
  - embroidery kit consumes embroidery floss, which is made from thread and dye
  - found in clothing stores, among hobby supplies, and occasionally in bedrooms
- added a new rare-ish material, lace
  - currently only used in one embroidery task, but can be broken down for lots of thread
  - found occasionally in clothing and hobby shops
- crafting recipes for embroidery floss and lace, and disassembly recipes to match
- added several new clothing mods, mostly themed around cosmetic personalization of clothes with fairly steep penalties to combat-readiness.
  - embroidery kit is used for personalization mods, which mostly just make things fancier
- lots of new utility mods for the tailor kit

I tried to make all these balanced and realistic, but please have a look if you're doing a deep review and let me know if they don't make sense. I can tweak as required or move anything that's a hard no to an optional game mod.

The new features are documented in JSON_INFO.md and in relevant source code.

#### Describe alternatives you've considered
- creating clothing mods that add item flags directly using `flag` field: this is extremely broken and a bad idea, and besides is limited to one flag
- there was no support for most of the other features, so that had to be done in C++

#### Testing
- extensively tested new clothing mods and interactions between them using debug stuff, fixed relevant UI stuff to display changes, ensured all new mod capabilities worked as intended
- some light spot-checking in gdb to make sure nothing obviously crazy was going on under the hood
- ran `make check` - but there didn't seem to be any clothing mod related unit tests (I can try my  hand at writing a few, but without a pre-existing baseline I don't know if I broke anything)
- travis CI is reporting some unit test failures to do with mutations, but I'm told those are not my fault
- ran astyle & json formatter (I set up a vim config to automate this, but did spot check manually just in case)

All testing done in curses mode, built with clang on arch linux (x86_64). Not tested in Mac/Android/Windows or with tiles. I can test with tiles, but don't have the other things.

#### Additional context
These enhancements will hopefully lay the groundwork for more interesting clothing mods in the future, and add a bunch of fun and useful things for tailors to do with their tool kits.

I have thoughts on more items and clothing mods, like a leatherworking toolkit that lets you add studs and other embellishments to leather gear; would like to work with @Rail-Runner on that if they're willing as they make progress on the tailoring overhaul.

I also have a few other ideas for this that will require further infrastructure, and some follow-up tasks to improve QoL especially on long duration mod tasks (see thread).

#### Screenshots:
Let's make this evening gown a little more practical:
![Screenshot from 2020-04-05 02-06-49](https://user-images.githubusercontent.com/523439/78471625-8e96ec80-76e7-11ea-8a4c-5f07b23dd31f.png)

This sling pack would be way cooler with some decorative embroidery. Also, the strap seems like a perfect place for a thermometer:
![Screenshot from 2020-04-05 02-07-55](https://user-images.githubusercontent.com/523439/78471642-aff7d880-76e7-11ea-9a7e-f8490c592dd9.png)

New clothing mod UI:
![Screenshot from 2020-04-05 02-08-50](https://user-images.githubusercontent.com/523439/78471654-c867f300-76e7-11ea-8cd3-ef1b5d6b187e.png)

nphyx commented:
I was trying not to be overly ambitious with this initial bit ... but I'm thinking, mods could further restrict by body part covered, material, and other parameters, which would enable them to be much more easily cross-compatible with other mods (as opposed to the restricted and valid_mods properties).

They could suppress as well as apply flags, which would enable refitting as a mod instead of a weird side-effect of repair/reinforce.

They could add or remove covered parts, which would let you have "sleeveless" as a mod instead of a crafting recipe.

One other feature I wanted to do but decided not to was the ability to damage items when removing a mod. This would make more complex modifications (e.g. the cutwork lace in the example mod) require repairs afterward, representing the patches you'd have to replace to restore the item's original function. I wasn't sure that was a good approach.

Anyway, any takers on any of this?

anothersimulacrum commented on src/clothing_mod.h:
```diff
@@ -6,6 +6,7 @@
 #include <string>
 #include <vector>
 #include <array>
+#include <set>
```
Why are you including set?  You don't appear to use it.

nphyx replied:
Good catch, I was using set for apply_flags and exclude_flags, but switched to vector later for easier interplay with item.cpp. I'll remove this in next commit.

anothersimulacrum commented on src/item.cpp:
```diff
@@ -5288,7 +5297,12 @@ int item::get_coverage() const
+    result += get_clothing_mod_val( clothing_mod_type_coverage );
+
+    // note: this may not be the best place to put this check?
+    return std::max( 0, std::min( result, 100 ) );
```
Why the min? Why is encumbrance capped at 100?

nphyx replied:
This is coverage. Since clothing_mod now has clothing_mod_type_coverage it's possible to create a mod that results in coverage < 0 or > 100, which is obviously wrong. Came up while testing, applying a coverage increase to a coat that already had 95% coverage.

I did not cap other values, though I wondered if having negative warmth, encumbrance, or protection should be maxed to 0 too. What happens if something has a negative final value here?

The question is whether putting it here in get_coverage() is the right way to handle it. Does anything bypass get_coverage() when doing coverage checks and access the value directly? since in the past nothing changed coverage dynamically? If so would it be better to fix those spots to always call get_coverage(), or better to handle the problem in a different way?

anothersimulacrum replied:
Ack, my bad!

I'd say the other values should be floored at 0 as well, and this is probably the right place to stick it. If other places are accessing it directly, they should probably be converted to use this.

nphyx replied:
Ok, I'll go ahead and do that since I'm already in there.

nphyx replied:
Ah yeah I probably should have checked if there was a clamp function >.< Thanks. My day job is mostly javascript, which doesn't have clamp in the core library so I'm in the habit of doing this manually <.<

LaVeyanFiend commented:
Why is this a mod and not mainline content?

nphyx commented:
> Why is this a mod and not mainline content?

Parts of the mod could be mainlined, but I saw there was a WIP on tailoring and wanted to wait on progress on that since it looks like tailoring materials and tools are going to change.

Other parts I think undermine game balance intent - while doing homework on this I found a convo somewhere where kevin specifically bounced the idea of having a mainline clothing mod that adds stylish because it's a perk intended for clothing with little to no protection. Sorry I don't have a link for that.

Meanwhile I figured making it optional would be reasonable (especially the stuff in Fashionista, which is mostly related to style and personalization).

Amoebka commented:
Fanciness of clothing has lost any balance value long ago. Morale gain from Stylish is capped and it's trivial to max it out via 0% coverage jewelry. You no longer make a choice between protection and looks.

nphyx commented:
> Fanciness of clothing has lost any balance value long ago. Morale gain from Stylish is capped and it's trivial to max it out via 0% coverage jewelry. You no longer make a choice between protection and looks.

Yeah I kinda noticed that too :/ I'm open to putting the embroidery stuff mainline if relevant people approve.

Unrelated: I'm not sure what these failures in the travis build log are, and don't know what they have to do with my changes, any pointers?

https://travis-ci.org/github/CleverRaven/Cataclysm-DDA/jobs/670548170?utm_medium=notification&utm_source=github_status

```
mutation_test.cpp:38: FAILED:
  CHECK( mutation_attempts > 0 )
with expansion:
  0 > 0
with messages:
  Current mutations: MANA_REGEN2 ROT2 THRESH_MANA MANA_CRYST_MANA FLIMSY2
  MANA_ADD3 ALBINO WEAK MANA_LUM MANA_SILENT_SPELL MANA_SEEKER_BOLTS
  MANA_SUBTLE_SPELL 
  Mutating towards MANA_REGEN3
```

```
mutation_test.cpp:122: FAILED:
  CHECK( breach_chance <= BREACH_CHANCE_MAX )
with expansion:
  0.53097f <= 0.4f
with message:
  MUTATIONS: MANA_MULT3 MANA_REGEN2 MANA_ADD3 MANA_SEEKER_BOLTS ALBINO WEAK
  FLIMSY2 MANA_CRYST_MANA ROT2 MANA_LUM 
```

https://travis-ci.org/github/CleverRaven/Cataclysm-DDA/jobs/670548172?utm_medium=notification&utm_source=github_status
```
0.110 s: npc_talk_conditionals
src/character.cpp:657:57: runtime error: division by zero
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior src/character.cpp:657:57 in 
```

anothersimulacrum commented:
Not your fault, no need for you to worry about it.

LaVeyanFiend commented:
Is it possible to modify the storage amounts through this, for a sewn pocket mod?

nphyx commented:
> Is it possible to modify the storage amounts through this, for an added pocket mod?

Changing storage was already possible, but there are no clothing mods that actually do it. I was going to throw it in there but it seemed like something that should be on the tailor kit, not the embroidery kit - and then I found out that items aren't generic_factory yet so it would mean replacing the tailor kit entirely in the mod, which I didn't like...

I could put it in mainline instead. Might be kinda broken to be able to stick pockets on your socks, glove liner, boonie hat, etc. though. The storage struggle would ... cease to be real.

Also I think I saw something in the works about changing up how pockets work, to do with nested containers.

nphyx commented:
> It's perfectly fit for vanilla. There's absolutely no reason to keep this content in a mod.

Looks like there's a strong consensus on this so I'll go ahead and move it to mainline :) I'll update the PR once it's done. If anything seems out of whack or just not suited for the core game I can pull that part back out and make it a mod.

I had a couple more thoughts on this today. I'm going to to add in some time and difficulty parameters. If this is going to be mainline I don't think it's realistic to, for example, embroider an entire jacket or evening gown in ~4 1/2 minutes (30 seconds * 9 pieces of material). Nor should a novice be able to do elaborate cutwork if they happen to get their hands on some lace (which already has a skill requirement to craft).

Given that, I'd like to tie in the use of the craft item/system so longer modification tasks can be paused or interrupted, but I'm not sure if that's going to be possible. I'll look at it, but if anybody has any tips or can tag in someone who is familiar with that system I'd appreciate it.

anothersimulacrum commented:
@ifreund is familiar with the crafting system, given that they wrote the whole thing for pausable crafts.

nphyx commented:
Dug into that a bit, and it looks like the system requires a recipe in order to generate a craft. So if I could dynamically create a recipe instance that consumed the original item and requisite materials and whose result was the same item with the mod applied, and then passed that into the crafting system, that might work?

Is anything going to explode if I try that? E.g. would that information all get serialized in a save, so that if you save and reload the craft will still know what it's supposed to make? I'll look into it myself later if nobody has time, I understand y'all are busy with your own things.

In any case this looks like it will be a little complicated so if I do it, it'll be a separate PR. For now I'm just going to make those time & difficulty parameters. I'll make them default to the previous hard-coded values so nothing else will change.

ifreund commented:
Currently all that is serialized with regards to the recipe is an id, which is then used to look up the recipe in the static data that ships with the game. So yeah, dynamically creating and serializing recipes/leveraging the crafting system isn't trivial and definitely belongs in a follow up PR.

Feel free to reach out on irc (`#CataclysmDDA` on `irc.freenode.net`) or [discord](https://discord.gg/jFEc7Yp) if you'd like to discuss implementation details.

nphyx commented:
Poked around in the source more, and it looks like a player_activity would be good middle ground solution.

Switching to the player_activity system is still going to be a bigger change than I had intended here though, so I'm going to go with the basic version I've got so far and do the activity thing as a follow-up PR if this one is approved (though I'll likely work on it anyway).

Meanwhile sitting around for 5 in-game hours helpless while your character works on a complex clothing mod doesn't feel great, so I'm going to keep the durations short. Do you guys use issues for follow-up todos? Should I file one if this accepted?

ifreund commented:
Yeah an issue would be perfect for that. You may want to have a look at the new activity infrastructure that just got merged #37450.

nphyx commented:
Ok, I've updated with the changes we discussed, and updated the description in the OP. I think I'm good with features on this for now but happy to take any more review/suggestions.

Sorry I-Am-Erk, I didn't intentionally tag you for review on this. Not sure why it did that.

anothersimulacrum commented:
You touched an SUS file, which caused him to be called for a review. Don't worry about it, he set it up so he's always called for review when someone touches those.

nphyx commented:
Ok, I had some new ideas that required a couple more features. OP updated to reflect. In brief:

- mods can target or exclude body parts
- mods can suppress flags on gear
- mods can have a maximum coverage value
- mods can have explicit material requirement amounts
- some new utility mods for tailor kit (@LaVeyanFiend pockets for leg and torso items!)

Stepping away from this for reals now, I have something else I want to work on. :D

ifreund commented on src/clothing_mod.h:
```diff
@@ -45,14 +49,84 @@ struct clothing_mod {
+     *
+     * @return bool true if the mod applies flags
+     */
+    bool applies_flags( ) const;
```
nit: these shouldn't have spaces (though i guess astyle doesn't enforce this)
```suggestion
    bool applies_flags() const;
```

ifreund commented on src/clothing_mod.h:
```diff
@@ -45,14 +49,84 @@ struct clothing_mod {
+     *
+     * @return bool true if the mod applies flags
+     */
+    bool suppresses_flags( ) const;
```
```suggestion
    bool suppresses_flags() const;
```

ifreund commented on src/clothing_mod.h:
```diff
@@ -45,14 +49,84 @@ struct clothing_mod {
+     *
+     * Flags are incompatible if an item has one of the flags this mod applies,
+     * or if this mod excludes one of its flags.
+         *
+         * Body parts are compatible if the item covers at least one of the valid_parts
+         * and does not cover any of the invalid_parts.
+         *
```
Looks like the indent got messed up here somehow
```suggestion
     *
     * Body parts are compatible if the item covers at least one of the valid_parts
     * and does not cover any of the invalid_parts.
     *
```

ifreund commented on src/iuse_actor.cpp:
```diff
@@ -4796,34 +4904,48 @@ int sew_advanced_actor::use( player &p, item &it, bool, const tripoint & ) const
-    comps.push_back( item_comp( repair_item, items_needed ) );
-    p.moves -= to_moves<int>( 30_seconds * p.fine_detail_vision_mod() );
-    p.practice( used_skill, items_needed * 3 + 3 );
+    int skill_level = p.get_skill_level( used_skill );
```
```suggestion
    const int skill_level = p.get_skill_level( used_skill );
```

Wishbringer commented on src/clothing_mod.h:
```diff
@@ -12,10 +12,11 @@
-class player;
+class JsonObject;
 class item;
+class player;
```
This was unused so it was removed, no need to add it back.

nphyx replied:
My bad, I'll get it fixed.

LaVeyanFiend commented on data/json/clothing_mods.json:
```diff
@@ -28,30 +161,143 @@
+    "mod_value": [ { "type": "encumbrance", "value": -2, "round_up": false }, { "type": "warmth", "value": -5, "round_up": false } ]
+  },
+  {
+    "type": "clothing_mod",
+    "id": "tailored_fingerless",
+    "flag": "tailored_fingerless",
```
Should have applied flag of ALLOWS_NATURAL_ATTACKS, as the crafted fingerless gloves do.

nphyx replied:
Nice thank you.

nphyx commented on data/json/clothing_mods.json:
```diff
@@ -28,30 +161,143 @@
+    "difficulty": 1,
+    "no_material_scaling": true,
+    "item_quantity": 12,
+    "mod_value": [ { "type": "encumbrance", "value": 3, "round_up": true }, { "type": "warmth", "value": 10, "round_up": true } ]
```
Good catch on the coverage mod.

I capped long sleeved at 70% because that seemed to be the upper limit of coverage for items that weren't already explicitly long-sleeved, based on a very unscientific process of spawning a few items I could think of and checking <.<

But it could be dialed in better if you have suggestions. I figure 100% arms coverage = it must already have long sleeves.

dissociativity commented:
I'm looking forward to this as something that'll allow more gear versatility instead of more or less making survivor gear the be all and end all of gear more or less for a huge portion of the game

nphyx commented:
> I'm looking forward to this as something that'll allow more gear versatility instead of more or less making survivor gear the be all and end all of gear more or less for a huge portion of the game

I wouldn't expect any kind of modded clothing to directly compete with the top-tier crafted armors, but this should give you more knobs for dialing in a piecemeal set to suit your needs. In particular you'll have more options for tuning warmth, coverage, and encumbrance vs. armor values.

Hopefully the code changes will also open the door for new mods that wouldn't have been possible before, especially in combination with Rail-Runner's work overhauling tailor crafts and the new recipes and materials that are coming with it.

It would for example allow someone to make a mod that integrates a respirator with your favorite mask, or changes the clothing layer that a piece rests on, or tailor any gear to fit mutants with unusually sized bodies - none of which I've done here, but it could be done with a couple quick JSON additions now instead of a bunch of crafting recipes. Some of that might not be a good fit for the base game but a mod focused on style over substance could definitely bring it in.

esotericist commented:
> I wouldn't expect any kind of modded clothing to directly compete with the top-tier crafted armors, but this should give you more knobs for dialing in a piecemeal set to suit your needs. In particular you'll have more options for tuning warmth, coverage, and encumbrance vs. armor values.

i'm actually of the opinion that crafted armors shouldn't actually be "top-tier" in their niches. survivor armor as we currently have it in-game should be something achieved with modifications, not a mysterious 'it's super awesome because i made it out of junk' endgame thing

the "top-tier" stuff for most categories of purpose should generally be pre-cataclysm (and thus not crafted) stuff, which *might* sometimes be customizable with modifications.

nphyx commented:
> survivor armor as we currently have it in-game should be something achieved with modifications, not a mysterious 'it's super awesome because i made it out of junk' endgame thing

I like the idea of survivor armor being something you achieve with mods rather than a specific crafting recipe. Would have to do further work on the difficulty and skill requirements with some target in mind for how many mods someone should be able to get away with, and/or deal with making some mods harder to achieve or integrate with others.

If we dialed in the balance and versatility of clothing & armor mods we could just kill the survivor gear line completely. Though it does look pretty cool in some tilesets and I'd hate to lose that.

> the "top-tier" stuff for most categories of purpose should generally be pre-cataclysm (and thus not crafted) stuff, which _might_ sometimes be customizable with modifications.

I think it would depend a lot on what you're optimizing for. Modern armor is built for modern combat scenarios, not scavenging in a world full of bitey, scratchy undead monsters, so it's not necessarily providing the right tradeoffs. A ballistic vest should bring nothing to the table when tangling with zeds, who will happily bite your arms off or throw you through a wall if they can't get at your chewy center and use the vest as a convenient grip while they're doing it. You'd definitely want it when dealing with remnant military and psycho survivors though.

Anyway the fact that everybody seems to use survivor stuff is a pretty good clue that it's overtuned, or at least that there aren't other viable options.

esotericist commented:
> Modern armor is built for modern combat scenarios, not scavenging in a world full of bitey, scratchy undead monsters, so it's not necessarily providing the right tradeoffs.

i was thinking of riot gear as my exemplar there. while this category of equipment obviously wasn't designed with things like hulks in mind, it's typically pretty well suited to helping mitigate low-velocity impact and cut damage. i have trouble believing that for defensive purposes, someone can beat that with scavenged materials.

to be clear: my assertion is that *best in a specific category* should be pre-cataclysm stuff. my position is that if you have something niche that straddles categories, then you should get there by modifying pre-cataclysm gear, or mixing pre-cataclysm gear. you are not going to get best-in-class for a specific category using scrap.

the thing i've been fighting against is the assumption that crafted gear can be 'best' at specific purposes, simply because it's "end-game" or "custom equipment" and the like.

nphyx commented:
> i was thinking of riot gear as my exemplar there. while this category of equipment obviously wasn't designed with things like hulks in mind, it's typically pretty well suited to helping mitigate low-velocity impact and cut damage. i have trouble believing that for defensive purposes, someone can beat that with scavenged materials.
> 
> to be clear: my assertion is that _best in a specific category_ should be pre-cataclysm stuff. my position is that if you have something niche that straddles categories, then you should get there by modifying pre-cataclysm gear, or mixing pre-cataclysm gear. you are not going to get best-in-class for a specific category using scrap.
> 
> the thing i've been fighting against is the assumption that crafted gear can be 'best' at specific purposes, simply because it's "end-game" or "custom equipment" and the like.

Yeah that makes sense. Maybe the equivalent of "heavy survivor gear" would be riot gear with added storage, some extra padding, and a little streamlining in the arms and legs to improve flexibility at the cost of defense.

kevingranade commented:
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/how-do-the-tailors-kit-clothing-mods-work-coding-wise/23337/2

dissociativity commented:
the problem survivor gear has to me, is it's objectively better than most 'better' armour due to how coverage ends up working, you're better off with heavy survivor gear, than fancy high end military armour as a stray bullet could miss your fancy vest very easily, but will always hit your survivor armour, same with other attacks, coverage makes anything below 100% significantly worse than armour without 100% coverage.

You can have armour that should be 'better' that ends up always being worse.

nphyx commented:
Yah because of that being able to change coverage may be a significant balance issue in this PR. It's way outside the scope here to "fix" it - the question is whether allowing players to mod coverage is going to upset balance.

Probably the real issue is that coverage as in "how much skin is covered" and coverage as in "how much of that area is actually armored" should be separate concerns. There's no way a survivor suit should be giving 100% *armor* coverage, given it's fashioned out of duct tape and random kevlar patches. But you would expect whatever warmth and environmental protection it provides to not have any gaps.

Anyway all that is above my pay grade so to speak and I suspect would be challenging to implement. I'd guess something like that gets looked at whenever the game moves away from hit point-based systems and toward individually modeled wounds.

ZhilkinSerg commented:
Sometimes no messages appear in the log after clothing was modified (or failed to be modified).

nphyx commented:
> Sometimes no messages appear in the log after clothing was modified (or failed to be modified).

Thanks for testing! I'll take a look.

LaVeyanFiend commented:
How plausible would an XL-ing option that adds the OVERSIZE flag be? We could then obsolete all of the seperately defined XL items & open up mutant clothing without bloating the .json

nphyx commented:
> How plausible would an "XL-ify"(name pending :P) option that adds the OVERSIZE flag be? We could then obsolete most/all of the seperately defined XL items & open up mutant clothing without bloating the recipe list & .json
> (Probably default on with applicable blacklisting so you can prevent just XLing things like gas masks which would need to be rebuilt to seal properly on huge mutants)

Should be possible right now. I even put a note about how to do it in the update to the JSON documents. Main thing would be to add `OVERSIZE` and suppress any other size modifier flags. My only concern is I don't know how much of "you can't wear armor" is an intentional balance thing vs. just not wanting to do a ton of duplicate oversized gear in the crafting system. We could easily exclude items with another flag if we specifically don't want them to be modified for mutants.

Wishbringer commented on src/iuse_actor.cpp:
```diff
@@ -4719,40 +4737,131 @@ int sew_advanced_actor::use( player &p, item &it, bool, const tripoint & ) const
+        if( mod.bash_resist() != temp_item.bash_resist() ) {
+            desc += format_desc_string( _( "Bash" ), mod.bash_resist(), temp_item.bash_resist(), true );
+        }
+        if( mod.bash_resist() != temp_item.bash_resist() ) {
```
```suggestion
        if( mod.cut_resist() != temp_item.cut_resist() ) {
```

LaVeyanFiend commented:
The pockets modification will have to be rewritten now that nested containers is merged.

Shodan14 commented:
> 
> 
> Closing per discord discussion - kevin is looking for something much more comprehensive on this than I'm willing/able to commit to (or possibly dropping the existing system entirely and making it work more like gun/tool mods - which is interesting).

It's such a shame to just let all this work go to waste :( Can't you just make it a mod at least until kevin's dreams materialize? @nphyx

nphyx commented:
> It's such a shame to just let all this work go to waste :( Can't you just make it a mod at least until kevin's dreams materialize? @nphyx

No, this requires C++ support and kevin doesn't want the required changes because he feels it will open the door to game balance issues in the core game regardless of whether the content is in a mod.
